## Combine pickled data into one dataframe to then use for datviz

In [89]:
import pandas as pd
import numpy as np

### Unpickle file with all the features

In [10]:
unpickled_feat_df = pd.read_pickle("./feat_dummy.pkl")
unpickled_feat_df

RestaurantName  population    income  risk  day_1  \
0                       STARBUCKS COFFEE        1682  100377.0     2      0   
1                    KELLY'S CAJUN GRILL        1682  100377.0     3      0   
2                             MONK'S PUB        1682  100377.0     3      0   
3                  GIORDANO'S ON JACKSON        1682  100377.0     3      0   
4                           Skadden Cafe        1682  100377.0     3      0   
...                                  ...         ...       ...   ...    ...   
72901                  BIRRIERIA OCOTLAN       96288   35534.0     3      1   
15512                              PROST       65474   68324.0     3      0   
63876  Benchwarmers Sports Bar and Grill       76015   38487.0     3      0   
33688                 THE FLAT TOP GRILL       47513   51294.0     3      0   
67842                             SUBWAY       87827   38210.0     3      1   

       day_3  day_2  day_4  day_0  month_6  ...  year_2019  year_2018  \
0          0      0      1      0        0  ...          0          0   
1          0      0      1      0        0  ...          0          0   
2          0      0      0      1        0  ...          0          0   
3          0      0      1      0        0  ...          0          0   
4          0      0      1      0        0  ...          0          0   
...      ...    ...    ...    ...      ...  ...        ...        ...   
72901      0      0      0      0        0  ...          0          0   
15512      1      0      0      0        0  ...          0          1   
63876      1      0      0      0        0  ...          0          0   
33688      1      0      0      0        1  ...          0          0   
67842      0      0      0      0        0  ...          0          0   

       month_9  year_2010  month_12  year_2013  year_2014  \
0            0          0         0          0          0   
1            0          0         0          0          1   
2            0          0         0          1          0   
3            0          0         0          0          0   
4            0          0         0          0          0   
...        ...        ...       ...        ...        ...   
72901        0          0         0          0          0   
15512        0          0         0          0          0   
63876        0          0         0          0          0   
33688        0          1         0          0          0   
67842        0          0         0          0          0   

       inspection_type_License  inspection_type_canvass  \
0                            0                        1   
1                            0                        1   
2                            0                        1   
3                            0                        0   
4                            0                        1   
...                        ...                      ...   
72901                        0                        1   
15512                        0                        0   
63876                        0                        1   
33688                        0                        0   
67842                        0                        1   

       inspection_type_Complaint_Other  
0                                    0  
1                                    0  
2                                    0  
3                                    0  
4                                    0  
...                                ...  
72901                                0  
15512                                1  
63876                                0  
33688                                0  
67842                                0  

[171396 rows x 34 columns]

#### Reset unpickled_feat_df

In [11]:
feat_df_reset = unpickled_feat_df.reset_index()
feat_df_reset

index                     RestaurantName  population    income  risk  \
0           0                   STARBUCKS COFFEE        1682  100377.0     2   
1           1                KELLY'S CAJUN GRILL        1682  100377.0     3   
2           2                         MONK'S PUB        1682  100377.0     3   
3           3              GIORDANO'S ON JACKSON        1682  100377.0     3   
4           4                       Skadden Cafe        1682  100377.0     3   
...       ...                                ...         ...       ...   ...   
171391  72901                  BIRRIERIA OCOTLAN       96288   35534.0     3   
171392  15512                              PROST       65474   68324.0     3   
171393  63876  Benchwarmers Sports Bar and Grill       76015   38487.0     3   
171394  33688                 THE FLAT TOP GRILL       47513   51294.0     3   
171395  67842                             SUBWAY       87827   38210.0     3   

        day_1  day_3  day_2  day_4  day_0  ...  year_2019  year_2018  month_9  \
0           0      0      0      1      0  ...          0          0        0   
1           0      0      0      1      0  ...          0          0        0   
2           0      0      0      0      1  ...          0          0        0   
3           0      0      0      1      0  ...          0          0        0   
4           0      0      0      1      0  ...          0          0        0   
...       ...    ...    ...    ...    ...  ...        ...        ...      ...   
171391      1      0      0      0      0  ...          0          0        0   
171392      0      1      0      0      0  ...          0          1        0   
171393      0      1      0      0      0  ...          0          0        0   
171394      0      1      0      0      0  ...          0          0        0   
171395      1      0      0      0      0  ...          0          0        0   

        year_2010  month_12  year_2013  year_2014  inspection_type_License  \
0               0         0          0          0                        0   
1               0         0          0          1                        0   
2               0         0          1          0                        0   
3               0         0          0          0                        0   
4               0         0          0          0                        0   
...           ...       ...        ...        ...                      ...   
171391          0         0          0          0                        0   
171392          0         0          0          0                        0   
171393          0         0          0          0                        0   
171394          1         0          0          0                        0   
171395          0         0          0          0                        0   

        inspection_type_canvass  inspection_type_Complaint_Other  
0                             1                                0  
1                             1                                0  
2                             1                                0  
3                             0                                0  
4                             1                                0  
...                         ...                              ...  
171391                        1                                0  
171392                        0                                1  
171393                        1                                0  
171394                        0                                0  
171395                        1                                0  

[171396 rows x 35 columns]

### Unpickle file with y (Results), Predictions, and Class Probabilities

In [5]:
unpickled_pred_df = pd.read_pickle("./pred_dummy.pkl")
unpickled_pred_df.head()

RestaurantName  Results_Pass_Fail  Prediction  \
0                        STARBUCKS COFFEE                  1           1   
1                     KELLY'S CAJUN GRILL                  1           1   
2                              MONK'S PUB                  1           1   
3                   GIORDANO'S ON JACKSON                  1           1   
4                            Skadden Cafe                  1           0   
...                                   ...                ...         ...   
171391                  BIRRIERIA OCOTLAN                  0           0   
171392                              PROST                  0           0   
171393  Benchwarmers Sports Bar and Grill                  0           0   
171394                 THE FLAT TOP GRILL                  0           0   
171395                             SUBWAY                  0           0   

                              Class Probabilities  
0                                    [0.05, 0.95]  
1                                      [0.0, 1.0]  
2                                      [0.0, 1.0]  
3                                  [0.025, 0.975]  
4        [0.5038690476190476, 0.4961309523809523]  
...                                           ...  
171391  [0.5425992063492063, 0.45740079365079367]  
171392   [0.7657142857142858, 0.2342857142857143]  
171393  [0.8568001443001443, 0.14319985569985566]  
171394   [0.8080952380952379, 0.1919047619047619]  
171395  [0.8159836160571455, 0.18401638394285455]  

[171396 rows x 4 columns]

### Drop 'index' and 'RestaurantName' columns from feat_df_reset 

In [12]:
df_to_concat = feat_df_reset.drop(columns=['index','RestaurantName'])
df_to_concat

population    income  risk  day_1  day_3  day_2  day_4  day_0  \
0             1682  100377.0     2      0      0      0      1      0   
1             1682  100377.0     3      0      0      0      1      0   
2             1682  100377.0     3      0      0      0      0      1   
3             1682  100377.0     3      0      0      0      1      0   
4             1682  100377.0     3      0      0      0      1      0   
...            ...       ...   ...    ...    ...    ...    ...    ...   
171391       96288   35534.0     3      1      0      0      0      0   
171392       65474   68324.0     3      0      1      0      0      0   
171393       76015   38487.0     3      0      1      0      0      0   
171394       47513   51294.0     3      0      1      0      0      0   
171395       87827   38210.0     3      1      0      0      0      0   

        month_6  month_3  ...  year_2019  year_2018  month_9  year_2010  \
0             0        0  ...          0          0        0          0   
1             0        0  ...          0          0        0          0   
2             0        0  ...          0          0        0          0   
3             0        0  ...          0          0        0          0   
4             0        0  ...          0          0        0          0   
...         ...      ...  ...        ...        ...      ...        ...   
171391        0        0  ...          0          0        0          0   
171392        0        0  ...          0          1        0          0   
171393        0        0  ...          0          0        0          0   
171394        1        0  ...          0          0        0          1   
171395        0        0  ...          0          0        0          0   

        month_12  year_2013  year_2014  inspection_type_License  \
0              0          0          0                        0   
1              0          0          1                        0   
2              0          1          0                        0   
3              0          0          0                        0   
4              0          0          0                        0   
...          ...        ...        ...                      ...   
171391         0          0          0                        0   
171392         0          0          0                        0   
171393         0          0          0                        0   
171394         0          0          0                        0   
171395         0          0          0                        0   

        inspection_type_canvass  inspection_type_Complaint_Other  
0                             1                                0  
1                             1                                0  
2                             1                                0  
3                             0                                0  
4                             1                                0  
...                         ...                              ...  
171391                        1                                0  
171392                        0                                1  
171393                        1                                0  
171394                        0                                0  
171395                        1                                0  

[171396 rows x 33 columns]

### Concatenate feat_df_reset with unpickled_pred_df


In [15]:
restaurant_result = pd.concat([unpickled_pred_df, df_to_concat], axis=1, sort=False)
restaurant_result

RestaurantName  Results_Pass_Fail  Prediction  \
0                        STARBUCKS COFFEE                  1           1   
1                     KELLY'S CAJUN GRILL                  1           1   
2                              MONK'S PUB                  1           1   
3                   GIORDANO'S ON JACKSON                  1           1   
4                            Skadden Cafe                  1           0   
...                                   ...                ...         ...   
171391                  BIRRIERIA OCOTLAN                  0           0   
171392                              PROST                  0           0   
171393  Benchwarmers Sports Bar and Grill                  0           0   
171394                 THE FLAT TOP GRILL                  0           0   
171395                             SUBWAY                  0           0   

                              Class Probabilities  population    income  risk  \
0                                    [0.05, 0.95]        1682  100377.0     2   
1                                      [0.0, 1.0]        1682  100377.0     3   
2                                      [0.0, 1.0]        1682  100377.0     3   
3                                  [0.025, 0.975]        1682  100377.0     3   
4        [0.5038690476190476, 0.4961309523809523]        1682  100377.0     3   
...                                           ...         ...       ...   ...   
171391  [0.5425992063492063, 0.45740079365079367]       96288   35534.0     3   
171392   [0.7657142857142858, 0.2342857142857143]       65474   68324.0     3   
171393  [0.8568001443001443, 0.14319985569985566]       76015   38487.0     3   
171394   [0.8080952380952379, 0.1919047619047619]       47513   51294.0     3   
171395  [0.8159836160571455, 0.18401638394285455]       87827   38210.0     3   

        day_1  day_3  day_2  ...  year_2019  year_2018  month_9  year_2010  \
0           0      0      0  ...          0          0        0          0   
1           0      0      0  ...          0          0        0          0   
2           0      0      0  ...          0          0        0          0   
3           0      0      0  ...          0          0        0          0   
4           0      0      0  ...          0          0        0          0   
...       ...    ...    ...  ...        ...        ...      ...        ...   
171391      1      0      0  ...          0          0        0          0   
171392      0      1      0  ...          0          1        0          0   
171393      0      1      0  ...          0          0        0          0   
171394      0      1      0  ...          0          0        0          1   
171395      1      0      0  ...          0          0        0          0   

        month_12  year_2013  year_2014  inspection_type_License  \
0              0          0          0                        0   
1              0          0          1                        0   
2              0          1          0                        0   
3              0          0          0                        0   
4              0          0          0                        0   
...          ...        ...        ...                      ...   
171391         0          0          0                        0   
171392         0          0          0                        0   
171393         0          0          0                        0   
171394         0          0          0                        0   
171395         0          0          0                        0   

        inspection_type_canvass  inspection_type_Complaint_Other  
0                             1                                0  
1                             1                                0  
2                             1                                0  
3                             0                                0  
4                             1                                0

### Pickle and export as csv files

In [16]:
pd.to_pickle(restaurant_result, "./restaurant_result_dataviz.pkl")

In [18]:
restaurant_result.to_csv('./restaurant_result_dataviz.csv')

### Read in final pickled_df

In [19]:
restaurant_result_dataviz_df = pd.read_pickle("./restaurant_result_dataviz.pkl")
restaurant_result_dataviz_df.head()

RestaurantName  Results_Pass_Fail  Prediction  \
0       STARBUCKS COFFEE                  1           1   
1    KELLY'S CAJUN GRILL                  1           1   
2             MONK'S PUB                  1           1   
3  GIORDANO'S ON JACKSON                  1           1   
4           Skadden Cafe                  1           0   

                        Class Probabilities  population    income  risk  \
0                              [0.05, 0.95]        1682  100377.0     2   
1                                [0.0, 1.0]        1682  100377.0     3   
2                                [0.0, 1.0]        1682  100377.0     3   
3                            [0.025, 0.975]        1682  100377.0     3   
4  [0.5038690476190476, 0.4961309523809523]        1682  100377.0     3   

   day_1  day_3  day_2  ...  year_2019  year_2018  month_9  year_2010  \
0      0      0      0  ...          0          0        0          0   
1      0      0      0  ...          0          0        0          0   
2      0      0      0  ...          0          0        0          0   
3      0      0      0  ...          0          0        0          0   
4      0      0      0  ...          0          0        0          0   

   month_12  year_2013  year_2014  inspection_type_License  \
0         0          0          0                        0   
1         0          0          1                        0   
2         0          1          0                        0   
3         0          0          0                        0   
4         0          0          0                        0   

   inspection_type_canvass  inspection_type_Complaint_Other  
0                        1                                0  
1                        1                                0  
2                        1                                0  
3                        0                                0  
4                        1                                0  

[5 rows x 37 columns]

### Show Predicted Counts and Results_Counts (0:Fail, 1:Pass)

In [24]:
Predicted_Counts = restaurant_result_dataviz_df.Prediction.value_counts()
Predicted_Counts

0    96836
1    74560
Name: Prediction, dtype: int64

In [25]:
Results_Counts = restaurant_result_dataviz_df.Results_Pass_Fail.value_counts()
Results_Counts

1    85698
0    85698
Name: Results_Pass_Fail, dtype: int64

### Observations

- Model was **88.5% correct** on restaurants **failing**
- Model was **87% correct** on restaurants **passing**

## CMatrix - TP, FN, FP, TN

In [26]:
def new_col(y_actual, y_pred):
    if y_actual == 1:
        if y_pred == 1:
            return 'TP'
        if y_pred == 0:
            return "FN"
    if y_actual == 0:
        if y_pred == 1:
            return 'FP'
        if y_pred == 0:
            return "TN"        

In [27]:
restaurant_result_dataviz_df['CMatrix'] = restaurant_result_dataviz_df.apply\
(lambda x: new_col(x['Results_Pass_Fail'],x['Prediction']), axis=1)

In [31]:
CM_df = restaurant_result_dataviz_df['CMatrix']
CM_df.to_frame()

CMatrix
0           TP
1           TP
2           TP
3           TP
4           FN
...        ...
171391      TN
171392      TN
171393      TN
171394      TN
171395      TN

[171396 rows x 1 columns]

### CMatrix vs. 'income', 'population', 'risk' columns

In [59]:
CMatrix_inc_pop_risk = restaurant_result_dataviz_df.groupby('CMatrix')['income','population','risk'].describe()
CMatrix_inc_pop_risk = CMatrix_inc_pop_risk.reset_index()
CMatrix_inc_pop_risk

CMatrix   income                                                         \
             count          mean           std      min      25%      50%   
0      FN  17687.0  45358.149714  15590.693719  14205.0  34293.0  40972.0   
1      FP   6549.0  48425.798290  17151.814217  14205.0  35283.0  41355.0   
2      TN  79149.0  43101.349455  14960.386118  14205.0  32466.0  40165.0   
3      TP  68011.0  45494.603917  16940.432096  14205.0  33631.0  40972.0   

                     population  ...                        risk            \
       75%       max      count  ...      75%       max    count      mean   
0  55316.0  100377.0    17687.0  ...  85771.0  113984.0  17687.0  2.887149   
1  55647.0  100377.0     6549.0  ...  87577.0  113984.0   6549.0  2.942434   
2  51294.0  100377.0    79149.0  ...  85771.0  113984.0  79149.0  2.790901   
3  55316.0  100377.0    68011.0  ...  79469.0  113984.0  68011.0  2.772919   

                                      
        std  min  25%  50%  75%  max  
0  0.323489  1.0  3.0  3.0  3.0  3.0  
1  0.233593  1.0  3.0  3.0  3.0  3.0  
2  0.437731  1.0  3.0  3.0  3.0  3.0  
3  0.442935  1.0  3.0  3.0  3.0  3.0  

[4 rows x 25 columns]

### CMatrix vs. 'income' column

In [60]:
CMatrix_inc = restaurant_result_dataviz_df.groupby('CMatrix')['income'].describe().sort_values(by='mean')
CMatrix_inc

count          mean           std      min      25%      50%  \
CMatrix                                                                   
TN       79149.0  43101.349455  14960.386118  14205.0  32466.0  40165.0   
FN       17687.0  45358.149714  15590.693719  14205.0  34293.0  40972.0   
TP       68011.0  45494.603917  16940.432096  14205.0  33631.0  40972.0   
FP        6549.0  48425.798290  17151.814217  14205.0  35283.0  41355.0   

             75%       max  
CMatrix                     
TN       51294.0  100377.0  
FN       55316.0  100377.0  
TP       55316.0  100377.0  
FP       55647.0  100377.0

### describe() 'income' column

In [61]:
income_desc = (restaurant_result_dataviz_df.income.describe().to_frame()).T
income_desc

count          mean           std      min      25%      50%  \
income  171396.0  44487.341373  15984.664129  14205.0  32622.0  40279.0   

            75%       max  
income  54698.0  100377.0

### Append income_desc and CMatrix_inc

In [63]:
CM_income = income_desc.append(CMatrix_inc)
CM_income = CM_income.reset_index()
CM_income

index     count          mean           std      min      25%      50%  \
0  income  171396.0  44487.341373  15984.664129  14205.0  32622.0  40279.0   
1      TN   79149.0  43101.349455  14960.386118  14205.0  32466.0  40165.0   
2      FN   17687.0  45358.149714  15590.693719  14205.0  34293.0  40972.0   
3      TP   68011.0  45494.603917  16940.432096  14205.0  33631.0  40972.0   
4      FP    6549.0  48425.798290  17151.814217  14205.0  35283.0  41355.0   

       75%       max  
0  54698.0  100377.0  
1  51294.0  100377.0  
2  55316.0  100377.0  
3  55316.0  100377.0  
4  55647.0  100377.0

### Difference in CMatrix and income

In [51]:
CM_inc_diff = restaurant_result_dataviz_df.groupby('CMatrix').income.describe()\
- restaurant_result_dataviz_df.income.describe()
CM_inc_diff = CM_inc_diff.reset_index()
CM_inc_diff

CMatrix     count         mean          std  min     25%     50%     75%  \
0      FN -153709.0   870.808341  -393.970410  0.0  1671.0   693.0   618.0   
1      FP -164847.0  3938.456917  1167.150088  0.0  2661.0  1076.0   949.0   
2      TN  -92247.0 -1385.991918 -1024.278010  0.0  -156.0  -114.0 -3404.0   
3      TP -103385.0  1007.262544   955.767967  0.0  1009.0   693.0   618.0   

   max  
0  0.0  
1  0.0  
2  0.0  
3  0.0

### Pickle CM_income and CM_inc_diff

In [52]:
CM_income.to_csv('./CM_inc_dataviz.csv')

In [53]:
CM_inc_diff.to_csv('./CM_inc_diff_dataviz.csv')

### Story to tell from this:

- **Average** income: 44487.34


- True Positive = Actually Passed & Predicted to Pass: TP difference = 1007.262544
- False Negative = Actually Passed & Predicted to Fail: FN difference = 870.808341
- False Positive = Actually Failed & Predicted to Pass: FP difference = 3938.456917
- True Negative = Actually Failed & Predicted to Fail: TN difference = -1385.991918


- For the False Negatives (FN): model predicted average income neighborhoods would fail their inspection when they actually passed
    - FN difference = 870.808341
- For the False Positives (FP): model predicted much higher income neighborhoods would pass their inspection when they actually failed
    - FP difference = 3938.456917


- The model has a strong bias towards higher income neighborhoods passing and a slight bias towards average income neighborhoods failing their health inspection

### CMatrix vs. 'population' column

In [69]:
CMatrix_population = restaurant_result_dataviz_df.groupby('CMatrix')['population'].describe().sort_values(by='mean')
CMatrix_population

count          mean           std   min      25%      50%      75%  \
CMatrix                                                                         
TP       68011.0  56484.812089  31900.841204  70.0  28832.0  57090.0  79469.0   
FP        6549.0  59458.545579  31197.533689  70.0  37990.0  65474.0  87577.0   
FN       17687.0  59671.657149  30075.211044  70.0  39155.0  65474.0  85771.0   
TN       79149.0  60259.599894  29457.279128  70.0  44197.0  65474.0  85771.0   

              max  
CMatrix            
TP       113984.0  
FP       113984.0  
FN       113984.0  
TN       113984.0

### describe() 'population' column

In [56]:
population_desc = (restaurant_result_dataviz_df.population.describe().to_frame()).T
population_desc

count          mean           std   min      25%      50%  \
population  171396.0  58670.460466  30630.495511  70.0  39155.0  65474.0   

                75%       max  
population  85771.0  113984.0

### Append population_desc and CMatrix_population

In [65]:
CM_pop = population_desc.append(CMatrix_population)
CM_pop = CM_pop.reset_index()
CM_pop

index     count          mean           std   min      25%      50%  \
0  population  171396.0  58670.460466  30630.495511  70.0  39155.0  65474.0   
1          TP   68011.0  56484.812089  31900.841204  70.0  28832.0  57090.0   
2          FP    6549.0  59458.545579  31197.533689  70.0  37990.0  65474.0   
3          FN   17687.0  59671.657149  30075.211044  70.0  39155.0  65474.0   
4          TN   79149.0  60259.599894  29457.279128  70.0  44197.0  65474.0   

       75%       max  
0  85771.0  113984.0  
1  79469.0  113984.0  
2  87577.0  113984.0  
3  85771.0  113984.0  
4  85771.0  113984.0

### Difference in CMatrix and population


In [70]:
CM_pop_diff = restaurant_result_dataviz_df.groupby('CMatrix').population.describe()\
- restaurant_result_dataviz_df.population.describe()
CM_pop_diff = CM_pop_diff.reset_index()
CM_pop_diff

CMatrix     count         mean          std  min      25%     50%     75%  \
0      FN -153709.0  1001.196683  -555.284467  0.0      0.0     0.0     0.0   
1      FP -164847.0   788.085114   567.038178  0.0  -1165.0     0.0  1806.0   
2      TN  -92247.0  1589.139428 -1173.216383  0.0   5042.0     0.0     0.0   
3      TP -103385.0 -2185.648377  1270.345693  0.0 -10323.0 -8384.0 -6302.0   

   max  
0  0.0  
1  0.0  
2  0.0  
3  0.0

### Pickle CM_pop and CM_pop_diff

In [71]:
CM_pop.to_csv('./CM_pop_dataviz.csv')

In [72]:
CM_pop_diff.to_csv('./CM_pop_diff_dataviz.csv')

### Story to tell from this:

- **Average** population: 58670.460466


- True Positive = Actually Passed & Predicted to Pass: TP difference = -2185.648377
- False Negative = Actually Passed & Predicted to Fail: FN difference = 1001.196683
- False Positive = Actually Failed & Predicted to Pass: FP difference = 788.085114
- True Negative = Actually Failed & Predicted to Fail: TN difference = 1589.139428


- For the False Negatives (FN): it predicted more populated neighborhoods would fail their inspection when they actually passed
    - FN difference = 1001.196683 
- Similarly, for the False Positives (FP): it predicted slightly more populated neighborhoods would pass their inspection when they actually failed
    - FP difference = 788.085114 

### CMatrix vs. 'risk' columm

In [73]:
CMatrix_risk = restaurant_result_dataviz_df.groupby('CMatrix')['risk'].describe().sort_values(by='mean')
CMatrix_risk

count      mean       std  min  25%  50%  75%  max
CMatrix                                                      
TP       68011.0  2.772919  0.442935  1.0  3.0  3.0  3.0  3.0
TN       79149.0  2.790901  0.437731  1.0  3.0  3.0  3.0  3.0
FN       17687.0  2.887149  0.323489  1.0  3.0  3.0  3.0  3.0
FP        6549.0  2.942434  0.233593  1.0  3.0  3.0  3.0  3.0

### describe() 'risk' column

In [74]:
risk_desc = (restaurant_result_dataviz_df.risk.describe().to_frame()).T
risk_desc

count      mean       std  min  25%  50%  75%  max
risk  171396.0  2.799488  0.425561  1.0  3.0  3.0  3.0  3.0

### Append risk_desc and CMatrix_risk

In [75]:
CM_risk = risk_desc.append(CMatrix_risk)
CM_risk = CM_risk.reset_index()
CM_risk

index     count      mean       std  min  25%  50%  75%  max
0  risk  171396.0  2.799488  0.425561  1.0  3.0  3.0  3.0  3.0
1    TP   68011.0  2.772919  0.442935  1.0  3.0  3.0  3.0  3.0
2    TN   79149.0  2.790901  0.437731  1.0  3.0  3.0  3.0  3.0
3    FN   17687.0  2.887149  0.323489  1.0  3.0  3.0  3.0  3.0
4    FP    6549.0  2.942434  0.233593  1.0  3.0  3.0  3.0  3.0

### Difference in CMatrix and 'risk'


In [76]:
CM_risk_diff = restaurant_result_dataviz_df.groupby('CMatrix').risk.describe()\
- restaurant_result_dataviz_df.risk.describe()
CM_risk_diff = CM_risk_diff.reset_index()
CM_risk_diff

CMatrix     count      mean       std  min  25%  50%  75%  max
0      FN -153709.0  0.087661 -0.102072  0.0  0.0  0.0  0.0  0.0
1      FP -164847.0  0.142946 -0.191967  0.0  0.0  0.0  0.0  0.0
2      TN  -92247.0 -0.008587  0.012170  0.0  0.0  0.0  0.0  0.0
3      TP -103385.0 -0.026569  0.017374  0.0  0.0  0.0  0.0  0.0

### Pickle CM_risk and CM_risk_diff

In [77]:
CM_risk.to_csv('./CM_risk_dataviz.csv')

In [78]:
CM_risk_diff.to_csv('./CM_risk_diff_dataviz.csv')

#### Story to tell from this:

- **Average** risk: 2.799488


- True Positive = Actually Passed & Predicted to Pass: TP difference = -0.026569
- False Negative = Actually Passed & Predicted to Fail: FN difference = 0.087661
- False Positive = Actually Failed & Predicted to Pass: FP difference = 0.142946
- True Negative = Actually Failed & Predicted to Fail: TN difference = -0.008587


- For the False Negatives (FN): it predicted basically the same risk restaurants would fail their inspection when they actually passed
    - FN difference = 0.087661
- For the False Positives (FP): it predicted restaurants with slightly higher risk would fail their inspection when they actually passed
    - FP difference = 0.142946


- So, the model has a slight bias towards higher risk restaurants failing.

## Combine dataframes for plotting

In [83]:
CM_risk

index     count      mean       std  min  25%  50%  75%  max
0  risk  171396.0  2.799488  0.425561  1.0  3.0  3.0  3.0  3.0
1    TP   68011.0  2.772919  0.442935  1.0  3.0  3.0  3.0  3.0
2    TN   79149.0  2.790901  0.437731  1.0  3.0  3.0  3.0  3.0
3    FN   17687.0  2.887149  0.323489  1.0  3.0  3.0  3.0  3.0
4    FP    6549.0  2.942434  0.233593  1.0  3.0  3.0  3.0  3.0

In [84]:
CM_pop

index     count          mean           std   min      25%      50%  \
0  population  171396.0  58670.460466  30630.495511  70.0  39155.0  65474.0   
1          TP   68011.0  56484.812089  31900.841204  70.0  28832.0  57090.0   
2          FP    6549.0  59458.545579  31197.533689  70.0  37990.0  65474.0   
3          FN   17687.0  59671.657149  30075.211044  70.0  39155.0  65474.0   
4          TN   79149.0  60259.599894  29457.279128  70.0  44197.0  65474.0   

       75%       max  
0  85771.0  113984.0  
1  79469.0  113984.0  
2  87577.0  113984.0  
3  85771.0  113984.0  
4  85771.0  113984.0

In [85]:
CM_income

index     count          mean           std      min      25%      50%  \
0  income  171396.0  44487.341373  15984.664129  14205.0  32622.0  40279.0   
1      TN   79149.0  43101.349455  14960.386118  14205.0  32466.0  40165.0   
2      FN   17687.0  45358.149714  15590.693719  14205.0  34293.0  40972.0   
3      TP   68011.0  45494.603917  16940.432096  14205.0  33631.0  40972.0   
4      FP    6549.0  48425.798290  17151.814217  14205.0  35283.0  41355.0   

       75%       max  
0  54698.0  100377.0  
1  51294.0  100377.0  
2  55316.0  100377.0  
3  55316.0  100377.0  
4  55647.0  100377.0

In [86]:
join_df = pd.concat([CM_pop, CM_income, CM_risk])
join_df

index     count          mean           std      min      25%  \
0  population  171396.0  58670.460466  30630.495511     70.0  39155.0   
1          TP   68011.0  56484.812089  31900.841204     70.0  28832.0   
2          FP    6549.0  59458.545579  31197.533689     70.0  37990.0   
3          FN   17687.0  59671.657149  30075.211044     70.0  39155.0   
4          TN   79149.0  60259.599894  29457.279128     70.0  44197.0   
0      income  171396.0  44487.341373  15984.664129  14205.0  32622.0   
1          TN   79149.0  43101.349455  14960.386118  14205.0  32466.0   
2          FN   17687.0  45358.149714  15590.693719  14205.0  34293.0   
3          TP   68011.0  45494.603917  16940.432096  14205.0  33631.0   
4          FP    6549.0  48425.798290  17151.814217  14205.0  35283.0   
0        risk  171396.0      2.799488      0.425561      1.0      3.0   
1          TP   68011.0      2.772919      0.442935      1.0      3.0   
2          TN   79149.0      2.790901      0.437731      1.0      3.0   
3          FN   17687.0      2.887149      0.323489      1.0      3.0   
4          FP    6549.0      2.942434      0.233593      1.0      3.0   

       50%      75%       max  
0  65474.0  85771.0  113984.0  
1  57090.0  79469.0  113984.0  
2  65474.0  87577.0  113984.0  
3  65474.0  85771.0  113984.0  
4  65474.0  85771.0  113984.0  
0  40279.0  54698.0  100377.0  
1  40165.0  51294.0  100377.0  
2  40972.0  55316.0  100377.0  
3  40972.0  55316.0  100377.0  
4  41355.0  55647.0  100377.0  
0      3.0      3.0       3.0  
1      3.0      3.0       3.0  
2      3.0      3.0       3.0  
3      3.0      3.0       3.0  
4      3.0      3.0       3.0

### Export join_df to csv to 

In [189]:
join_df.to_csv('./join_df_dataviz.csv')

## Plotting with plotly

### Import necessary modules

In [92]:
import plotly.figure_factory as ff
import plotly.express as px
import plotly.graph_objs as go
from plotly.offline import iplot

### Create new column to apply colors for each feature in join_df

In [93]:
join_new_col = ['Population', 'Population', 'Population', 'Population', 'Population',\
                'Income', 'Income','Income','Income','Income',\
               'Risk', 'Risk', 'Risk', 'Risk', 'Risk']

In [94]:
join_df['Feature'] = join_new_col

In [95]:
join_df

index     count          mean           std      min      25%  \
0  population  171396.0  58670.460466  30630.495511     70.0  39155.0   
1          TP   68011.0  56484.812089  31900.841204     70.0  28832.0   
2          FP    6549.0  59458.545579  31197.533689     70.0  37990.0   
3          FN   17687.0  59671.657149  30075.211044     70.0  39155.0   
4          TN   79149.0  60259.599894  29457.279128     70.0  44197.0   
0      income  171396.0  44487.341373  15984.664129  14205.0  32622.0   
1          TN   79149.0  43101.349455  14960.386118  14205.0  32466.0   
2          FN   17687.0  45358.149714  15590.693719  14205.0  34293.0   
3          TP   68011.0  45494.603917  16940.432096  14205.0  33631.0   
4          FP    6549.0  48425.798290  17151.814217  14205.0  35283.0   
0        risk  171396.0      2.799488      0.425561      1.0      3.0   
1          TP   68011.0      2.772919      0.442935      1.0      3.0   
2          TN   79149.0      2.790901      0.437731      1.0      3.0   
3          FN   17687.0      2.887149      0.323489      1.0      3.0   
4          FP    6549.0      2.942434      0.233593      1.0      3.0   

       50%      75%       max     Feature  
0  65474.0  85771.0  113984.0  Population  
1  57090.0  79469.0  113984.0  Population  
2  65474.0  87577.0  113984.0  Population  
3  65474.0  85771.0  113984.0  Population  
4  65474.0  85771.0  113984.0  Population  
0  40279.0  54698.0  100377.0      Income  
1  40165.0  51294.0  100377.0      Income  
2  40972.0  55316.0  100377.0      Income  
3  40972.0  55316.0  100377.0      Income  
4  41355.0  55647.0  100377.0      Income  
0      3.0      3.0       3.0        Risk  
1      3.0      3.0       3.0        Risk  
2      3.0      3.0       3.0        Risk  
3      3.0      3.0       3.0        Risk  
4      3.0      3.0       3.0        Risk

### Scatter Plots

In [96]:
fig = px.scatter(join_df, x='index', y='mean', color='Feature')
fig.show()

### 3D plots

In [104]:
CM_income

index     count          mean           std      min      25%      50%  \
0  income  171396.0  44487.341373  15984.664129  14205.0  32622.0  40279.0   
1      TN   79149.0  43101.349455  14960.386118  14205.0  32466.0  40165.0   
2      FN   17687.0  45358.149714  15590.693719  14205.0  34293.0  40972.0   
3      TP   68011.0  45494.603917  16940.432096  14205.0  33631.0  40972.0   
4      FP    6549.0  48425.798290  17151.814217  14205.0  35283.0  41355.0   

       75%       max  
0  54698.0  100377.0  
1  51294.0  100377.0  
2  55316.0  100377.0  
3  55316.0  100377.0  
4  55647.0  100377.0

In [107]:
fig = px.scatter_3d(CM_income, x="index", y="mean", z="std", \
                    color="mean", size="mean", hover_name="index",\
                    color_discrete_map = {"Joly": "blue", "Bergeron": "green", "Coderre":"red"})
fig.show()

In [108]:
CM_pop

index     count          mean           std   min      25%      50%  \
0  population  171396.0  58670.460466  30630.495511  70.0  39155.0  65474.0   
1          TP   68011.0  56484.812089  31900.841204  70.0  28832.0  57090.0   
2          FP    6549.0  59458.545579  31197.533689  70.0  37990.0  65474.0   
3          FN   17687.0  59671.657149  30075.211044  70.0  39155.0  65474.0   
4          TN   79149.0  60259.599894  29457.279128  70.0  44197.0  65474.0   

       75%       max  
0  85771.0  113984.0  
1  79469.0  113984.0  
2  87577.0  113984.0  
3  85771.0  113984.0  
4  85771.0  113984.0

In [110]:
fig = px.scatter_3d(CM_pop, x="index", y="mean", z="std", \
                    color="mean", size="mean", hover_name="index",\
                    color_discrete_map = {"Joly": "blue", "Bergeron": "green", "Coderre":"red"})
fig.show()

In [111]:
CM_risk

index     count      mean       std  min  25%  50%  75%  max
0  risk  171396.0  2.799488  0.425561  1.0  3.0  3.0  3.0  3.0
1    TP   68011.0  2.772919  0.442935  1.0  3.0  3.0  3.0  3.0
2    TN   79149.0  2.790901  0.437731  1.0  3.0  3.0  3.0  3.0
3    FN   17687.0  2.887149  0.323489  1.0  3.0  3.0  3.0  3.0
4    FP    6549.0  2.942434  0.233593  1.0  3.0  3.0  3.0  3.0

In [112]:
fig = px.scatter_3d(CM_risk, x="index", y="mean", z="std", \
                    color="mean", size="mean", hover_name="index",\
                    color_discrete_map = {"Joly": "blue", "Bergeron": "green", "Coderre":"red"})
fig.show()